scipy/stats/_distn_infrastructure.py

```diff
@@ -1634,7 +1635,29 @@ def _penalized_nnlf(self, theta, x):
             return inf
         x = asarray((x-loc) / scale)
         n_log_scale = len(x) * log(scale)
-        return self._nnlf_and_penalty(x, args) + n_log_scale
+        return self._nlff_and_penalty(x, args, self._logpxf) + n_log_scale
+
+    def _penalized_nlpsf(self, theta, x):
+        """Penalized negative log product spacing function.
+        i.e., - sum (log (diff (cdf (x, theta))), axis=0) + penalty
+        where theta are the parameters (including loc and scale)
+        Follows [1] of scipy.stats.fit
+        """
+        loc, scale, args = self._unpack_loc_scale(theta)
+        if not self._argcheck(*args) or scale <= 0:
+            return inf
+        x = (np.sort(x) - loc)/scale
+
+        def log_psf(x, *args):
+            # simplest possible implementation for starters
+            x, lj = np.unique(x, return_counts=True)
+            cdf_data = self._cdf(x, *args) if x.size else []
+            cdf = np.concatenate(([0], cdf_data, [1]))
```

Comment:
From Wikipedia:
> For convenience also denote x(0) = −∞ and x(n+1) = +∞.

The CDF at -∞ is 0 and at +∞ it's 1, hence the concatenation.

But Shao and Hahn's generalization (for repeated observations) only adds x(0) = −∞.

![image](https://user-images.githubusercontent.com/6570539/194236237-2ec1c719-252f-4723-90a7-5e81e56edf38.png)

I'm conflicted.

Reply:
The "spacing" in Maximum Spacing Estimation refers to the spacing between PDFs, that is, the difference between CDFs of points in x. For continuous distributions, we expect there to be some non-zero spacing between the CDFs of x(n) and x(n+1) = +∞, so it makes sense to add x(n+1) = +∞. For discrete distributions, though, it is not uncommon to get samples at the upper end of the support of a discrete distribution, at which the CDF is 1. But if we append x(n+1) = +∞ in this case, we get a "spacing" between the CDFs of x(n) and x(n+1) = +∞ that equals zero. This would mean that the product spacing of the true parameters is often 0 for discrete distribution. (Another way of looking at it: the negative log of this is the objective function to minimize, and it would be infinite for the true parameters.) We don't want that. So for discrete distributions, it is not appropriate to append x(n+1) = +∞.

 (When we take the log of these 0s that occur for the true parameters, we get the `invalid` `RuntimeWarning` reported [below](https://github.com/scipy/scipy/pull/17141#pullrequestreview-1132409801).)

Reply:
Oops I misread. The paper addresses how to handle this.